src/Compilers/CSharp/Portable/BoundTree/BoundDecisionDag.cs

```diff
@@ -0,0 +1,210 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    partial class BoundDecisionDag
+    {
+        public IEnumerable<BoundDecisionDag> Successors()
+        {
+            switch (this)
+            {
+                case BoundEvaluationPoint p:
+                    yield return p.Next;
+                    yield break;
+                case BoundDecisionPoint p:
+                    Debug.Assert(p.WhenFalse != null);
+                    yield return p.WhenFalse;
+                    Debug.Assert(p.WhenTrue != null);
+                    yield return p.WhenTrue;
+                    yield break;
+                case BoundDecision d:
+                    yield break;
+                case BoundWhenClause w:
+                    Debug.Assert(w.WhenTrue != null);
+                    yield return w.WhenTrue;
+                    if (w.WhenFalse != null)
+                    {
+                        yield return w.WhenFalse;
+                    }
+
+                    yield break;
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(this.Kind);
+            }
+        }
+
+        private HashSet<LabelSymbol> _reachableLabels;
+
+        public HashSet<LabelSymbol> ReachableLabels
+        {
+            get
+            {
+                if (_reachableLabels == null)
+                {
+                    // Compute the set of reachable labels at the BoundDecision leaves.  We do this
+                    // iteratively rather than recursively so that we do not overflow the stack on a
+                    // large switch statement.
+                    var workQueue = ArrayBuilder<BoundDecisionDag>.GetInstance();
+                    var processed = PooledHashSet<BoundDecisionDag>.GetInstance();
+                    var reachableLabels = new HashSet<LabelSymbol>();
+                    workQueue.Add(this);
+                    while (workQueue.Count != 0)
+                    {
+                        BoundDecisionDag node = workQueue.Pop();
+                        if (node == null || !processed.Add(node))
```

Comment:
It seems better to avoid adding nulls to the queue in the first place. There could be other side effects from adding unnecessarily, like causing a re-allocation and copying of the queue array.

Reply:
The only place null can be added is as the false branch of a when clause. In that case the null is the next thing processed, so they do not pile up in the work queue.

---
In reply to: [176225387](https://github.com/dotnet/roslyn/pull/24059#discussion_r176225387) [](ancestors = 176225387)